registry/storage/driver/gcs/gcs.go

```diff
@@ -517,68 +481,85 @@ func (w *writer) writeChunk() error {
 	}
 	// if their is no sessionURI yet, obtain one by starting the session
 	if w.sessionURI == "" {
-		w.sessionURI, err = startSession(w.client, w.bucket, w.name)
+		w.sessionURI, err = w.newSession()
 	}
 	if err != nil {
 		return err
 	}
-	nn, err := putChunk(w.client, w.sessionURI, w.buffer[0:chunkSize], w.offset, -1)
-	w.offset += nn
+	n, err := w.putChunk(ctx, w.sessionURI, w.buffer[0:chunkSize], w.offset, -1)
+	w.offset += n
 	if w.offset > w.size {
 		w.size = w.offset
 	}
 	// shift the remaining bytes to the start of the buffer
-	w.buffSize = copy(w.buffer, w.buffer[int(nn):w.buffSize])
+	w.buffSize = copy(w.buffer, w.buffer[int(n):w.buffSize])
 
 	return err
 }
 
 func (w *writer) Write(p []byte) (int, error) {
-	err := w.checkClosed()
-	if err != nil {
-		return 0, err
+	if w.closed {
+		return 0, fmt.Errorf("already closed")
+	} else if w.cancelled {
+		return 0, fmt.Errorf("already cancelled")
 	}
 
-	var nn int
-	for nn < len(p) {
-		n := copy(w.buffer[w.buffSize:], p[nn:])
+	var (
+		written int
+		err     error
+	)
+
+	for written < len(p) {
+		n := copy(w.buffer[w.buffSize:], p[written:])
 		w.buffSize += n
 		if w.buffSize == cap(w.buffer) {
-			err = w.writeChunk()
+			err = w.writeChunk(w.ctx)
 			if err != nil {
 				break
 			}
 		}
-		nn += n
+		written += n
 	}
 	w.size = w.offset + int64(w.buffSize)
-	return nn, err
+	return written, err
 }
 
 // Size returns the number of bytes written to this FileWriter.
 func (w *writer) Size() int64 {
 	return w.size
 }
 
-func (w *writer) init(path string) error {
-	res, err := getObject(w.client, w.bucket, w.name, 0)
+func (w *writer) init(ctx context.Context) error {
+	attrs, err := w.object.Attrs(ctx)
 	if err != nil {
 		return err
 	}
-	defer res.Body.Close()
-	if res.Header.Get("Content-Type") != uploadSessionContentType {
-		return storagedriver.PathNotFoundError{Path: path}
+
+	if attrs.ContentType != uploadSessionContentType {
+		return storagedriver.PathNotFoundError{Path: w.object.ObjectName()}
 	}
-	offset, err := strconv.ParseInt(res.Header.Get("X-Goog-Meta-Offset"), 10, 64)
+
+	offset, err := strconv.ParseInt(attrs.Metadata["Offset"], 10, 64)
 	if err != nil {
 		return err
 	}
-	buffer, err := io.ReadAll(res.Body)
+
+	r, err := w.object.NewReader(ctx)
 	if err != nil {
 		return err
 	}
-	w.sessionURI = res.Header.Get("X-Goog-Meta-Session-URI")
-	w.buffSize = copy(w.buffer, buffer)
+	defer r.Close()
+
+	for err == nil && w.buffSize < len(w.buffer) {
+		var n int
+		n, err = r.Read(w.buffer[w.buffSize:])
+		w.buffSize += n
+	}
+	if err != nil && err != io.EOF {
+		return err
+	}
+
+	w.sessionURI = attrs.Metadata["Session-URI"]
```

Comment:
Mostly trying to get my head around here; this was previously using `X-Goog-Meta-Session-URI`. Is this a header we're always setting ourselves, or does this come from elsewhere (and if so, must we continue looking for both `X-Goog-Meta-Session-URI` and `Session-URI` ?)

Reply:
I looked into this when I was reviewing: they are the same thing. Arbitrary application-defined object metadata is conveyed over the wire using HTTP headers named `X-Goog-Metadata-{{name}}`; the Google Cloud SDK abstracts away that detail.

Reply:
Ah, thanks!